### PR TITLE
feat: cluster relay + versioned persistence (v1.20.0)

### PR DESCRIPTION
**Pull request by pyshx**
## Summary

Two generic, reusable extension points for building **multi-node** Yjs servers on ygo — both transport-agnostic, with reference implementations and tests. Motivated by an in-house Go Y-WebSocket backend that needs Redis-backed cross-instance fan-out and a versioned cloud-persistence adapter; only the generic interfaces land here, the backend-specific transports stay in the backend.

### `cluster` package — cross-node fan-out (updates **and** awareness)
- `Relay`/`Sink` interfaces + an in-proc `MemRelay` reference impl.
- Wired into `provider/websocket` via `(*Server).AttachRelay`; the server publishes local doc updates **and** awareness to the relay and injects remote ones. Echo prevention rides the existing `origin` sentinel (pointer-identity, never crosses the wire).
- Closes the gap where the old doc-sample clustered adapter punted on cross-node **awareness** — it's now first-class via `KindAwareness`.
- Adds `(*Server).GetAwareness` / `(*Server).Rooms` so `*Server` satisfies `cluster.Sink` directly.

### `persistence` package — versioned, crash-safe persistence
- `VersionedPersistence` interface: `Load` / `AppendUpdate` / `ListVersions` / `GetUpdate` / `MaterializeAt` / `CaptureSnapshot` / `RestoreSnapshot` / `PruneAfter` / `Compact` / `Delete`.
- Memory + file reference impls.
- Exported `RunConformance(t, factory)` so downstream adapters are held to the same contract — including a **crash-safe `PruneAfter`** case (`target=0` and `target=2`) proving a mid-prune crash never resurrects a version `> target` (checkpoint acts as a hard read-ceiling).
- `LegacyAdapter` shim maps the new interface onto the existing `NewServerWithPersistence(LoadDoc/StoreUpdate)`.

**Backward compatible:** the existing `PersistenceAdapter` path and provider API are unchanged; all relay machinery is a no-op when no relay is attached.

## Test plan
- `go test ./... -race` green across `cluster`, `persistence`, `provider/websocket`.
- `golangci-lint run` clean.
- Conformance suite (incl. crash-safe `PruneAfter target=0`) runs on both memory and file impls (verified RUN, not skip; negative-mutation checked).
- No change to Yjs/yrs binary compatibility.

## Release
`v1.20.0` — `CHANGELOG.md` updated; new `docs/CLUSTERING.md`, updated `docs/PERSISTENCE.md`.

### CODE REVIEW

File: cluster/mem_relay.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		return ErrRelayNotStarted
+	}
+	targets := make([]*memNode, len(m.nodes))
+	copy(targets, m.nodes)
+	m.mu.Unlock()
+
+	in := Inbound{Room: out.Room, Kind: out.Kind, Data: out.Data}
+	for _, n := range targets {
+		select {
+		case n.ch <- in:
```
Comment (bnimit) [inline cluster/mem_relay.go:157]: This send can panic with `send on closed channel` if `Close()` runs after `Publish` has passed the closed-check above but hasn't reached this line yet. `n.ctx` is the Start-supplied caller ctx — `Close` doesn't cancel it — so the third select arm doesn't save us either. The current tests don't hit this because they `defer Close` after `wg.Wait`, but adversarial ordering will trip it.

Easiest fix: add a relay-level `done chan struct{}` closed by `Close` and select on it here. Happy to push a patch if it's useful.

File: provider/websocket/cluster.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+		rm, err := s.getOrCreateRoom(ctx, in.Room)
+		if err != nil {
+			return err
+		}
+		if err := crdt.ApplyUpdateV1(rm.doc, in.Data, s.relaySentinel); err != nil {
+			return err
+		}
+		// Rebroadcast to locally-connected peers. BroadcastUpdate re-validates
+		// and frames; it does not re-apply to the doc (already applied above).
+		return s.BroadcastUpdate(ctx, in.Room, in.Data)
```
Comment (bnimit) [inline provider/websocket/cluster.go]: `BroadcastUpdate` fires `s.OnInject` (inject.go:171-179). `OnInject` is documented as a hook for *server-side writes* — operators wire it up for rate limits, content policies, etc. If the hook rejects an inbound relay update, the local doc on L160 is already mutated but peers never see the change → silent divergence between this node and the rest of the cluster. The error also gets swallowed by `MemRelay` (`_ = sink.Inject(...)`).

Cleanest fix is probably an internal `broadcastUpdateNoHook` path that `Inject` uses; that keeps `OnInject` scoped to writes originating locally. Want me to draft it?

File: persistence/conformance.go [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+			if lr.Version != 0 {
+				t.Fatalf("post-prune(0) Load version = %d, want 0", lr.Version)
+			}
+			if got := textOf(t, lr.Update); got != "" {
+				t.Fatalf("post-prune(0) Load text = %q, want empty", got)
+			}
+		})
+	})
+
+	t.Run("PruneAfterCrashSafe_NoSpuriousFutureVersions", func(t *testing.T) {
```
Comment (bnimit) [inline persistence/conformance.go:477]: The `PruneAfterCrashSafe` test is great — and `target=0` is exactly the right boundary case. One small extension worth adding while we're here: after the simulated crash, do an `AppendUpdate` and assert the new version is at least `maxPrunedVersion + 1` (i.e. for `target=0`, expect v6+, not v1).

The checkpoint correctly hides the leaked files from readers today, but if a future refactor accidentally lets a subsequent `Append` reuse v1..v5 (silently overwriting leaked-but-still-on-disk files), no current test catches it. A 3-line assert post-restart locks the version-monotonicity invariant in.

File: cluster/mem_relay_test.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
+package cluster_test
```
Comment (bnimit) [inline cluster/mem_relay_test.go:1]: One test gap worth filling before merge: spin a few hundred publishers and call `Close` mid-flight. Catches the `send on closed channel` race I flagged inline on `mem_relay.go`.

And over in `provider/websocket/cluster_test.go`, a parallel test: assert that `OnInject` does NOT fire when a remote update arrives via `Inject` — catches the silent-divergence bug directly. Both are small (~30 lines each) and exercise paths the current suite doesn't reach.

**Review by bnimit — COMMENTED** [review]
Read through this end-to-end, ran the suite under `-race` (all green), and dug into the persistence crash-safety claim. Overall this looks like a really solid piece of work — the checkpoint-as-hard-ceiling pattern in `persistence/file.go` is correctly enforced in every read path I traced (`listUpdateVersions`, `GetUpdate`, `nextVersion`, `memRoom.visibleRecords`), the `target=0` boundary case is exactly the right test, and the sentinel-by-pointer-identity echo guard composes cleanly with the awareness path.

Two real bugs to flag (inline on the lines), plus a couple of small test gaps I think are worth filling before merge. Otherwise it's good to go — I'd be happy to ship this as v1.20.0 after these.

File: persistence/memory.go [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+	for _, rec := range r.records {
+		if rec.version <= target {
+			kept = append(kept, rec)
+		}
+	}
+	r.records = kept
+	if r.nextVer <= target {
+		r.nextVer = target + 1
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (pyshx) [reply persistence/memory.go:343]: Fixed in ab5f619. A successful `PruneAfter` now clears the checkpoint ceiling (`checkpointSet=false`, `rolledBack=nil`, `nextVer=target+1`) so later appends are visible again; the ceiling is kept only on the simulated-crash early return. `AppendUpdate` also finishes a crash-interrupted prune before appending. Added conformance coverage that an append after `PruneAfter` is visible.

File: provider/websocket/server.go [hunk 6]
```diff
@@ -458,6 +475,15 @@ func NewServer() *Server {
...
 	s.shutdownOnce.Do(func() { close(s.shutdownCh) })
 
+	// Cancel the relay delivery context so its goroutines wind down, then close
+	// the relay. No-op when no relay is attached.
+	if s.relayCancel != nil {
+		s.relayCancel()
+	}
+	if s.relay != nil {
+		_ = s.relay.Close()
+	}
```
[the remark this answers is not in the document]
Comment (pyshx) [reply provider/websocket/server.go]: Fixed in ab5f619. `Server.Shutdown` no longer calls `relay.Close()` — it only cancels this node's relay context. The caller owns the (commonly shared) relay's lifetime now; documented on `AttachRelay` and the `Shutdown` comment.

File: cluster/relay.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+type Relay interface {
+	// Publish broadcasts a locally-originated change to all other nodes. It is
+	// the caller's responsibility (the provider wiring) to drop changes whose
+	// Origin is the relay sentinel before calling Publish.
+	Publish(ctx context.Context, out Outbound) error
+	// Start binds the relay to a Sink and begins delivering inbound changes.
+	// It must be called exactly once before the relay delivers anything. The
+	// supplied ctx governs the relay's delivery lifetime; cancelling it (or
+	// calling Close) stops delivery.
+	Start(ctx context.Context, sink Sink) error
```
[the remark this answers is not in the document]
Comment (pyshx) [reply cluster/relay.go]: Fixed in ab5f619. Reworded `Relay.Start`: each node (each `Server`) calls `Start` once; a relay shared across nodes is `Start`-ed once per node (a shared `MemRelay` therefore sees multiple `Start` calls). Dropped the misleading "exactly once".

File: cluster/mem_relay.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+	"errors"
+	"sync"
+)
+
+// ErrRelayClosed is returned by MemRelay.Publish / Start after Close.
+var ErrRelayClosed = errors.New("cluster: relay closed")
+
+// ErrRelayNotStarted is returned by MemRelay.Publish before Start has bound a
+// Sink, or by Start when called a second time.
+var ErrRelayNotStarted = errors.New("cluster: relay not started")
```
[the remark this answers is not in the document]
Comment (pyshx) [reply cluster/mem_relay.go]: Fixed in ab5f619. Reworded: `ErrRelayNotStarted` is returned by `Publish` before any node has `Start`-ed (no Sink bound) and by `Start` when `sink` is nil. `MemRelay` supports multiple `Start` calls and does not reject subsequent ones. Removed the "called a second time" clause.

File: cluster/mem_relay.go [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+// MemRelay is an in-process, channel-backed Relay. It is the reference
+// implementation: multiple nodes (each a websocket.Server) share one MemRelay
+// instance; a change Published by one node is delivered to every OTHER node's
+// Sink. It is safe for concurrent use and primarily intended for tests and
+// single-process multi-server simulations.
```
[the remark this answers is not in the document]
Comment (pyshx) [reply cluster/mem_relay.go:23]: Fixed in ab5f619. The type comment now says a published change is delivered to **every** node's Sink (including the publisher, which drops its own change via the echo sentinel), matching `Publish`.

File: persistence/conformance.go [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+			if err != nil {
+				t.Fatalf("Load: %v", err)
+			}
+			if lr.Version > 2 {
+				t.Fatalf("Load head version %d > target 2 after prune", lr.Version)
+			}
+			if got := textOf(t, lr.Update); got != "ba" {
+				t.Fatalf("post-prune Load text = %q, want %q", got, "ba")
+			}
+		})
```
[the remark this answers is not in the document]
Comment (pyshx) [reply persistence/conformance.go:436]: Added in ab5f619. The `PruneAfter` subtest now appends after the prune and asserts the new version `> target`, that `Load`'s head advances to it, that `ListVersions` contains it, and that the materialized state reflects the fresh update — directly catching the freeze on both reference impls.

File: persistence/file.go [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+		if perr != nil {
+			continue
+		}
+		if Version(n) > target {
+			if rmErr := os.Remove(filepath.Join(f.updatesDir(room), e.Name())); rmErr != nil {
+				return rmErr
+			}
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (pyshx) [reply persistence/file.go:522]: Fixed in ab5f619. `FilePersistence.PruneAfter` removes the checkpoint file after a successful delete (kept only on the crash early-return), and `AppendUpdate` finishes a crash-interrupted prune (deletes leaked files + removes the checkpoint) before writing — so the room can advance instead of being frozen at `target`.

File: persistence/file.go [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+	}
+
+	// Step 1: write checkpoint (atomic). From this point readers clamp to target.
+	if err := f.writeCheckpoint(room, &checkpoint{target: target, rolledBack: rolledBack}); err != nil {
+		f.mu.Unlock()
+		return err
+	}
+	crash := f.crashAfterCheckpoint
+	f.mu.Unlock()
+
```
[the remark this answers is not in the document]
Comment (pyshx) [reply persistence/file.go]: Fixed in ab5f619. `PruneAfter` now holds `f.mu` for the entire sequence (checkpoint write → crash hook → deletes → checkpoint removal) with no intermediate unlock; the crash hook runs under the lock, mirroring `MemoryPersistence`.

File: provider/websocket/cluster.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+func (s *Server) AttachRelay(r cluster.Relay) error {
+	if r == nil {
+		return ErrNilRelay
+	}
+	var attachErr error
+	s.relayOnce.Do(func() {
+		s.relay = r
+		s.relaySentinel = new(struct{})
+		s.relayCtx, s.relayCancel = context.WithCancel(context.Background())
+		attachErr = r.Start(s.relayCtx, s)
```
[the remark this answers is not in the document]
Comment (pyshx) [reply provider/websocket/cluster.go]: Fixed in ab5f619. `AttachRelay` now calls `relay.Start` **before** committing any state; on a `Start` error it cancels the context and returns without setting `s.relay`, so the call is retryable. Replaced `sync.Once` with a mutex + an `s.relay != nil` check.

File: provider/websocket/cluster.go [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+	unsubDoc := r.doc.OnUpdate(func(update []byte, origin any) {
+		if origin == sentinel {
+			return // echo guard: this change arrived via the relay
+		}
+		// Copy the update: the slice handed to OnUpdate observers may alias
+		// internal buffers, and Publish must not block — transports are expected
+		// to enqueue and return, so the data may be read after this observer returns.
+		cp := append([]byte(nil), update...)
+		_ = s.relay.Publish(s.relayCtx, cluster.Outbound{
```
[the remark this answers is not in the document]
Comment (pyshx) [reply provider/websocket/cluster.go]: Fixed in ab5f619 along the lines you suggested. The provider now buffers outbound relay events on a bounded queue (cap 1024) drained by a dedicated worker. The `doc.OnUpdate` / `awareness.OnChange` observers do a non-blocking enqueue (drop + counter on sustained overflow), so the CRDT commit path never blocks regardless of the relay impl, and the worker logs `Publish` failures. `MemRelay` therefore stays lossless for tests (its blocking is now off the hot path).

File: cluster/mem_relay.go (hunk 1 above)
Comment (pyshx) [reply cluster/mem_relay.go:157]: Good catch — fixed in ab5f619. `Close` no longer closes the per-node channels; instead a relay-level `done chan struct{}` (selected on in both `Publish`'s send loop and each node's `run` loop) stops delivery. Since `n.ch` is never closed, the send can't panic. Added `TestMemRelay_CloseDuringPublish_NoPanic` (400 publishers × 50, `Close` mid-flight, `-race`) — it reproduces the `send on closed channel` panic on the pre-fix code and passes after. Thanks for the precise diagnosis and the `done`-channel suggestion.

File: provider/websocket/cluster.go (hunk 2 above)
Comment (pyshx) [reply provider/websocket/cluster.go]: Fixed in ab5f619 with the no-hook approach you suggested. `BroadcastUpdate` now delegates to `broadcastUpdate(ctx, room, update, fireHook bool)`; `Inject` calls it with `fireHook=false`, so a remote relay update (already `ApplyUpdateV1`'d into the local doc) is rebroadcast to peers **without** firing `OnInject` — which stays scoped to locally-originated writes — eliminating the silent-divergence path. Added `TestInteg_Inject_KindSync_DoesNotFireOnInject`, which also asserts `OnInject` still fires for a subsequent local `BroadcastUpdate`.

File: persistence/conformance.go (hunk 3 above)
Comment (pyshx) [reply persistence/conformance.go:477]: Added in ab5f619 — with one adjustment to the exact assertion worth flagging. The crash-safe subtest now appends after the simulated crash and asserts the recovered version is `> target` (and that `Load`'s content is state-at-`target` ⊕ the fresh update, with no leaked version resurfacing), rather than `>= maxPruned+1`. Both reference impls — and a Rust-clock-style adapter — legitimately **reuse** `target+1` after a prune (dense continuation; `FilePersistence` derives the next version from max-on-disk, which is `target`), so a strict-monotonic assertion would wrongly fail conformant stores. The silent-overwrite risk you raised is closed regardless: `AppendUpdate` finishes the interrupted prune (deletes the leaked files) **before** the reused version is written, and the test asserts `GetUpdate` for every leaked version other than the densely-reused one returns not-found. Happy to switch the library to a strict-monotonic contract instead if you'd prefer — let me know.

File: cluster/mem_relay_test.go (hunk 4 above)
Comment (pyshx) [reply cluster/mem_relay_test.go:1]: Both added in ab5f619: a `MemRelay` close-during-publish race stress test (`cluster/mem_relay_test.go`, guards the send-on-closed race) and a provider test (`provider/websocket/cluster_test.go`) asserting `OnInject` is **not** fired on a relay `Inject` while it still fires for a local `BroadcastUpdate`. Both run under `-race` and were verified to fail on the pre-fix code.
